Importing python libraries to be used:

In [1]:
import pandas as pd
import numpy as np

There are 8 different csv files regarding each year. 
<br>To make full analysis throughout 2015-2022, I am going to join each file into main dataframe using 'for in' loop. 

In [2]:
#names of csv source files
years = ['2015','2016','2017','2018','2019','2020','2021','2022']

In [3]:
main_df = pd.DataFrame(pd.read_csv(r"./source_files/2015.csv"))
main_df['Year'] = '2015'
for i in range(1, len(years)):
    sub_df = pd.read_csv(r"./source_files/"+years[i]+r".csv")
    sub_df['Year'] = years[i]
    main_df = pd.concat([main_df,sub_df],axis=0)
    main_df = main_df.reset_index(drop=True)
main_df.head()


Country          Region  Happiness Rank  Happiness Score  \
0  Switzerland  Western Europe             1.0            7.587   
1      Iceland  Western Europe             2.0            7.561   
2      Denmark  Western Europe             3.0            7.527   
3       Norway  Western Europe             4.0            7.522   
4       Canada   North America             5.0            7.427   

   Standard Error  Economy (GDP per Capita)   Family  \
0         0.03411                   1.39651  1.34951   
1         0.04884                   1.30232  1.40223   
2         0.03328                   1.32548  1.36058   
3         0.03880                   1.45900  1.33095   
4         0.03553                   1.32629  1.32261   

   Health (Life Expectancy)  Freedom  Trust (Government Corruption)  ...  \
0                   0.94143  0.66557                        0.41978  ...   
1                   0.94784  0.62877                        0.14145  ...   
2                   0.87464  0.64938                        0.48357  ...   
3                   0.88521  0.66973                        0.36503  ...   
4                   0.90563  0.63297                        0.32957  ...   

   Explained by: Freedom to make life choices  Explained by: Generosity  \
0                                         NaN                       NaN   
1                                         NaN                       NaN   
2                                         NaN                       NaN   
3                                         NaN                       NaN   
4                                         NaN                       NaN   

  Explained by: Perceptions of corruption  Dystopia + residual  RANK  \
0                                     NaN                  NaN   NaN   
1                                     NaN                  NaN   NaN   
2                                     NaN                  NaN   NaN   
3                                     NaN                  NaN   NaN   
4                                     NaN                  NaN   NaN   

   Happiness score  Whisker-high  Whisker-low  Dystopia (1.83) + residual  \
0              NaN           NaN          NaN                         NaN   
1              NaN           NaN          NaN                         NaN   
2              NaN           NaN          NaN                         NaN   
3              NaN           NaN          NaN                         NaN   
4              NaN           NaN          NaN                         NaN   

   Explained by: GDP per capita  
0                           NaN  
1                           NaN  
2                           NaN  
3                           NaN  
4                           NaN  

[5 rows x 57 columns]

Let's see what the dataframe looks like:

In [4]:
main_df.shape

(1231, 57)

The df has 1231 rows and 57 columns. 
<br>The number of columns isn't expected. This might mean that the attributes of each subset of data have different naming so they were added as separate columns.
<br>Let's have a look at the columns:

In [5]:
#how many columns each csv has
column_check = {}

column_check = {years[i] : pd.read_csv(r"./source_files/"+years[i]+r".csv").shape[1] for i in range(0, len(years))} #using 1st index of attribute 'shape' which is number of columns
print(column_check)


{'2015': 12, '2016': 13, '2017': 12, '2018': 9, '2019': 9, '2020': 20, '2021': 20, '2022': 12}


In [6]:
#how the columns are named
main_df.columns.sort_values()


Index(['Country', 'Country name', 'Country or region',
       'Dystopia (1.83) + residual', 'Dystopia + residual',
       'Dystopia Residual', 'Dystopia.Residual', 'Economy (GDP per Capita)',
       'Economy..GDP.per.Capita.',
       'Explained by: Freedom to make life choices',
       'Explained by: GDP per capita', 'Explained by: Generosity',
       'Explained by: Healthy life expectancy',
       'Explained by: Log GDP per capita',
       'Explained by: Perceptions of corruption',
       'Explained by: Social support', 'Family', 'Freedom',
       'Freedom to make life choices', 'Freedom to make life choices_',
       'GDP per capita', 'Generosity', 'Generosity_', 'Happiness Rank',
       'Happiness Score', 'Happiness score', 'Happiness.Rank',
       'Happiness.Score', 'Health (Life Expectancy)',
       'Health..Life.Expectancy.', 'Healthy life expectancy',
       'Healthy life expectancy_', 'Ladder score', 'Ladder score in Dystopia',
       'Logged GDP per capita', 'Lower Confidence 

There is significant redundancy in the columns.
<br>For example, information about countries is split into three columns:
* Country
* Country name
* Country or region
<br> The files have between 9-20 columns. Let's have deeper look into 2020 and 2021 as they have much more columns than the rest of subsets:


In [7]:

for i in range(5, 6): #(5,6) are indexes of years list: 2020 and 2021 
    df_20_21 = pd.read_csv(r"./source_files/"+years[i]+r".csv")
df_20_21.columns

Index(['Country name', 'Regional indicator', 'Ladder score',
       'Standard error of ladder score', 'upperwhisker', 'lowerwhisker',
       'Logged GDP per capita', 'Social support_', 'Healthy life expectancy_',
       'Freedom to make life choices_', 'Generosity_',
       'Perceptions of corruption_', 'Ladder score in Dystopia',
       'Explained by: Log GDP per capita', 'Explained by: Social support',
       'Explained by: Healthy life expectancy',
       'Explained by: Freedom to make life choices',
       'Explained by: Generosity', 'Explained by: Perceptions of corruption',
       'Dystopia + residual'],
      dtype='object')

As suspected, these years contain duplicated columns of six happiness factors. Let's compare ranges of these factors:

In [8]:
df_20_21[['Logged GDP per capita',
'Explained by: Log GDP per capita',
'Social support_',
'Explained by: Social support',
'Healthy life expectancy_',
 'Explained by: Healthy life expectancy',
 'Freedom to make life choices_',
 'Explained by: Freedom to make life choices',
 'Generosity_','Explained by: Generosity',
 'Perceptions of corruption_','Explained by: Perceptions of corruption']].describe()

Logged GDP per capita  Explained by: Log GDP per capita  \
count             153.000000                        153.000000   
mean                9.295706                          0.868771   
std                 1.201588                          0.372416   
min                 6.492642                          0.000000   
25%                 8.350645                          0.575862   
50%                 9.456313                          0.918549   
75%                10.265124                          1.169229   
max                11.450681                          1.536676   

       Social support_  Explained by: Social support  \
count       153.000000                    153.000000   
mean          0.808721                      1.155607   
std           0.121453                      0.286866   
min           0.319460                      0.000000   
25%           0.737217                      0.986718   
50%           0.829204                      1.203987   
75%           0.906747                      1.387139   
max           0.974670                      1.547567   

       Healthy life expectancy_  Explained by: Healthy life expectancy  \
count                153.000000                             153.000000   
mean                  64.445529                               0.692869   
std                    7.057848                               0.254094   
min                   45.200001                               0.000000   
25%                   58.961712                               0.495443   
50%                   66.305145                               0.759818   
75%                   69.289192                               0.867249   
max                   76.804581                               1.137814   

       Freedom to make life choices_  \
count                     153.000000   
mean                        0.783360   
std                         0.117786   
min                         0.396573   
25%                         0.714839   
50%                         0.799805   
75%                         0.877709   
max                         0.974998   

       Explained by: Freedom to make life choices  Generosity_  \
count                                  153.000000   153.000000   
mean                                     0.463583    -0.014568   
std                                      0.141172     0.151809   
min                                      0.000000    -0.300907   
25%                                      0.381457    -0.127015   
50%                                      0.483293    -0.033665   
75%                                      0.576665     0.085429   
max                                      0.693270     0.560664   

       Explained by: Generosity  Perceptions of corruption_  \
count                153.000000                  153.000000   
mean                   0.189375                    0.733120   
std                    0.100401                    0.175172   
min                    0.000000                    0.109784   
25%                    0.115006                    0.683019   
50%                    0.176745                    0.783122   
75%                    0.255510                    0.849151   
max                    0.569814                    0.935585   

       Explained by: Perceptions of corruption  
count                               153.000000  
mean                                  0.130718  
std                                   0.113097  
min                                   0.000000  
25%                                   0.055805  
50%                                   0.098435  
75%                                   0.163064  
max                                   0.533162

According to the report documentation, compared columns on the right are modified data used as final output in all other data files. To avoid discrepancies in data ranges, I will use columns to the right.
<br>Our final column adjustment looks as below:

In [9]:
correct_names = {
'Country name': 'Country',
'Country or region': 'Country',
'Dystopia (1.83) + residual': 'Dystopia Residual',
'Dystopia + residual': 'Dystopia Residual',
'Dystopia.Residual': 'Dystopia Residual',
'Economy (GDP per Capita)': 'GDP per Capita',
'Economy..GDP.per.Capita.': 'GDP per Capita',
'Explained by: Freedom to make life choices': 'Freedom to Make Life Choices',
'Explained by: GDP per capita': 'GDP per Capita',
'Explained by: Generosity': "Generosity",
'Explained by: Healthy life expectancy': 'Healthy Life Expectancy',
'Explained by: Log GDP per capita': 'GDP per Capita',
'Explained by: Perceptions of corruption' : 'Corruption Perception',
'Explained by: Social support': 'Social Support',
'Family': 'Social Support',
'Freedom': 'Freedom to Make Life Choices',
'Freedom to make life choices': 'Freedom to Make Life Choices',
'GDP per capita': 'GDP per Capita',
'Happiness score': 'Happiness Score',
'Happiness.Rank': 'Happiness Rank',
'Happiness.Score': 'Happiness Score',
'Health (Life Expectancy)': 'Healthy Life Expectancy',
'Health..Life.Expectancy.': 'Healthy Life Expectancy',
'Healthy life expectancy': 'Healthy Life Expectancy',
'Ladder score': 'Happiness Score',
'Overall rank': 'Happiness Rank',
'Perceptions of corruption': 'Corruption Perception',
'RANK': 'Happiness Rank',
'Regional indicator': 'Region',
'Score': 'Happiness Score',
'Social support': 'Social Support',
'Trust (Government Corruption)': 'Corruption Perception',
'Trust..Government.Corruption.': 'Corruption Perception'
}

The columns have finally equalized names. However, each year has them as separate columns which needs to be fixed.

In [10]:
main_df = main_df.rename(columns=correct_names)
main_df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1231 entries, 0 to 1230
Data columns (total 57 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Country                         617 non-null    object 
 1   Region                          315 non-null    object 
 2   Happiness Rank                  315 non-null    float64
 3   Happiness Score                 315 non-null    float64
 4   Standard Error                  158 non-null    float64
 5   GDP per Capita                  315 non-null    float64
 6   Social Support                  470 non-null    float64
 7   Healthy Life Expectancy         315 non-null    float64
 8   Freedom to Make Life Choices    470 non-null    float64
 9   Corruption Perception           315 non-null    float64
 10  Generosity                      782 non-null    float64
 11  Dystopia Residual               315 non-null    float64
 12  Year                            12

The easiest method would be using concatenating code used earlier with addition of renaming columns while joining.
<br>It's time to focus on data inside.
<br>
<br>Having looked at main_df's tail:
<br>#1 There is unnecessary 1230th row which needs to be dropped.
<br>#2 Some countries have unnecessary asterix (*) which could disrupt the data and needs to be removed.

In [11]:

main_df = pd.DataFrame(pd.read_csv(r"./source_files/2015.csv"))
main_df = main_df.rename(columns=correct_names)
main_df['Year'] = '2015'

for i in range(1, len(years)):
    sub_df = pd.read_csv(r"./source_files/"+years[i]+r".csv")
    sub_df = sub_df.rename(columns=correct_names)
    sub_df['Year'] = years[i]
    main_df = pd.concat([main_df,sub_df],axis=0)
    main_df = main_df.reset_index(drop=True)

main_df['Year'] = main_df['Year'].astype(int)
main_df #it allows us to see both head and tail of the dataframe


Country          Region  Happiness Rank  Happiness Score  \
0     Switzerland  Western Europe             1.0            7.587   
1         Iceland  Western Europe             2.0            7.561   
2         Denmark  Western Europe             3.0            7.527   
3          Norway  Western Europe             4.0            7.522   
4          Canada   North America             5.0            7.427   
...           ...             ...             ...              ...   
1226      Rwanda*             NaN           143.0            3.268   
1227     Zimbabwe             NaN           144.0            2.995   
1228      Lebanon             NaN           145.0            2.955   
1229  Afghanistan             NaN           146.0            2.404   
1230           xx             NaN           147.0              NaN   

      Standard Error  GDP per Capita  Social Support  Healthy Life Expectancy  \
0            0.03411         1.39651         1.34951                  0.94143   
1            0.04884         1.30232         1.40223                  0.94784   
2            0.03328         1.32548         1.36058                  0.87464   
3            0.03880         1.45900         1.33095                  0.88521   
4            0.03553         1.32629         1.32261                  0.90563   
...              ...             ...             ...                      ...   
1226             NaN         0.00100         0.00000                  0.46200   
1227             NaN         0.00100         0.00100                  0.27000   
1228             NaN         1.39200         0.00000                  0.63100   
1229             NaN         0.00100         0.00000                  0.28900   
1230             NaN             NaN             NaN                      NaN   

      Freedom to Make Life Choices  Corruption Perception  ...  lowerwhisker  \
0                          0.66557                0.41978  ...           NaN   
1                          0.62877                0.14145  ...           NaN   
2                          0.64938                0.48357  ...           NaN   
3                          0.66973                0.36503  ...           NaN   
4                          0.63297                0.32957  ...           NaN   
...                            ...                    ...  ...           ...   
1226                       0.62100                0.54400  ...           NaN   
1227                       0.32900                0.10500  ...           NaN   
1228                       0.10300                0.03400  ...           NaN   
1229                       0.00000                0.00500  ...           NaN   
1230                           NaN                    NaN  ...           NaN   

      Logged GDP per capita  Social support_  Healthy life expectancy_  \
0                       NaN              NaN                       NaN   
1                       NaN              NaN                       NaN   
2                       NaN              NaN                       NaN   
3                       NaN              NaN                       NaN   
4                       NaN              NaN                       NaN   
...                     ...              ...                       ...   
1226                    NaN              NaN                       NaN   
1227                    NaN              NaN                       NaN   
1228                    NaN              NaN                       NaN   
1229                    NaN              NaN                       NaN   
1230                    NaN              NaN                       NaN   

      Freedom to make life choices_  Generosity_  Perceptions of corruption_  \
0                               NaN          NaN                         NaN   
1                               NaN          NaN                         NaN   
2                               NaN          NaN                         NaN   
3                         

Number of columns is reduced to 29. 
<br>Having looked at main_df.info: 
<br>#1 I only need first 12 columns, with Years included. The rest can be omitted.
<br>#2 I won't need Standard Error column (4th index).
<br>#3 Region and Happiness Rank columns have null values.
* Happiness Rank can be easily calculated based on Happiness Score of the countries (which thankfully there's no null values) 
* Region could be retrieved based on data from other years (none of the countries have moved anywhere and changed their region, I suppose...)

In [12]:
main_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1231 entries, 0 to 1230
Data columns (total 29 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Country                         1231 non-null   object 
 1   Region                          617 non-null    object 
 2   Happiness Rank                  929 non-null    float64
 3   Happiness Score                 1230 non-null   float64
 4   Standard Error                  158 non-null    float64
 5   GDP per Capita                  1230 non-null   float64
 6   Social Support                  1230 non-null   float64
 7   Healthy Life Expectancy         1230 non-null   float64
 8   Freedom to Make Life Choices    1230 non-null   float64
 9   Corruption Perception           1229 non-null   float64
 10  Generosity                      1230 non-null   float64
 11  Dystopia Residual               918 non-null    float64
 12  Year                            12

In [13]:

#reducing columns - from 1st to 'Year'
cut_df = main_df.loc[:,:'Year']

#getting rid of Standard Error column and the last row
cut_df = cut_df.drop(columns='Standard Error')
cut_df = cut_df.drop(axis=0, index=1230)

#replacing null value by country Emirates mean of Corruption Perception column
emirates = cut_df[cut_df['Country'].str.contains('Emirates')]
cut_df['Corruption Perception'] = cut_df['Corruption Perception'].fillna(emirates['Corruption Perception'].mean())

#getting rid of asterix from countries' names
cut_df['Country'] = cut_df['Country'].str.replace('*','', regex=True)

cut_df.tail()

Country Region  Happiness Rank  Happiness Score  GDP per Capita  \
1225     Botswana    NaN           142.0            3.471           1.503   
1226       Rwanda    NaN           143.0            3.268           0.001   
1227     Zimbabwe    NaN           144.0            2.995           0.001   
1228      Lebanon    NaN           145.0            2.955           1.392   
1229  Afghanistan    NaN           146.0            2.404           0.001   

      Social Support  Healthy Life Expectancy  Freedom to Make Life Choices  \
1225           0.001                    0.280                         0.571   
1226           0.000                    0.462                         0.621   
1227           0.001                    0.270                         0.329   
1228           0.000                    0.631                         0.103   
1229           0.000                    0.289                         0.000   

      Corruption Perception  Generosity  Dystopia Residual  Year  
1225                  0.102       0.012              0.000  2022  
1226                  0.544       0.187              0.001  2022  
1227                  0.105       0.106              0.001  2022  
1228                  0.034       0.082              0.000  2022  
1229                  0.005       0.089              1.263  2022

Let's focus on null values on Region and Happiness Rank.
<br>We clearly see that 2017-2019 and 2022 are missing data on Region and 2020-2021 are missing data on Happiness Rank:

In [14]:
counted_by_year = cut_df.groupby('Year').count()
counted_by_year

Country  Region  Happiness Rank  Happiness Score  GDP per Capita  \
Year                                                                     
2015      158     158             158              158             158   
2016      157     157             157              157             157   
2017      155       0             155              155             155   
2018      156       0             156              156             156   
2019      156       0             156              156             156   
2020      153     153               0              153             153   
2021      149     149               0              149             149   
2022      146       0             146              146             146   

      Social Support  Healthy Life Expectancy  Freedom to Make Life Choices  \
Year                                                                          
2015             158                      158                           158   
2016             157                      157                           157   
2017             155                      155                           155   
2018             156                      156                           156   
2019             156                      156                           156   
2020             153                      153                           153   
2021             149                      149                           149   
2022             146                      146                           146   

      Corruption Perception  Generosity  Dystopia Residual  
Year                                                        
2015                    158         158                158  
2016                    157         157                157  
2017                    155         155                155  
2018                    156         156                  0  
2019                    156         156                  0  
2020                    153         153                153  
2021                    149         149                149  
2022                    146         146                146

In [15]:
#checking indexing of 2020-2021 years - 782:1083
missing_rank = cut_df[cut_df.Year.isin([2020,2021])]
missing_rank

Country              Region  Happiness Rank  Happiness Score  \
782       Finland      Western Europe             NaN           7.8087   
783       Denmark      Western Europe             NaN           7.6456   
784   Switzerland      Western Europe             NaN           7.5599   
785       Iceland      Western Europe             NaN           7.5045   
786        Norway      Western Europe             NaN           7.4880   
...           ...                 ...             ...              ...   
1079      Lesotho  Sub-Saharan Africa             NaN           3.5120   
1080     Botswana  Sub-Saharan Africa             NaN           3.4670   
1081       Rwanda  Sub-Saharan Africa             NaN           3.4150   
1082     Zimbabwe  Sub-Saharan Africa             NaN           3.1450   
1083  Afghanistan          South Asia             NaN           2.5230   

      GDP per Capita  Social Support  Healthy Life Expectancy  \
782         1.285190        1.499526                 0.961271   
783         1.326949        1.503449                 0.979333   
784         1.390774        1.472403                 1.040533   
785         1.326502        1.547567                 1.000843   
786         1.424207        1.495173                 1.008072   
...              ...             ...                      ...   
1079        0.451000        0.731000                 0.007000   
1080        1.099000        0.724000                 0.340000   
1081        0.364000        0.202000                 0.407000   
1082        0.457000        0.649000                 0.243000   
1083        0.370000        0.000000                 0.126000   

      Freedom to Make Life Choices  Corruption Perception  Generosity  \
782                       0.662317               0.477857    0.159670   
783                       0.665040               0.495260    0.242793   
784                       0.628954               0.407946    0.269056   
785                       0.661981               0.144541    0.362330   
786                       0.670201               0.434101    0.287985   
...                            ...                    ...         ...   
1079                      0.405000               0.015000    0.103000   
1080                      0.539000               0.088000    0.027000   
1081                      0.627000               0.493000    0.227000   
1082                      0.359000               0.075000    0.157000   
1083                      0.000000               0.010000    0.122000   

      Dystopia Residual  Year  
782            2.762835  2020  
783            2.432741  2020  
784            2.350267  2020  
785            2.460688  2020  
786            2.168266  2020  
...                 ...   ...  
1079           1.800000  2021  
1080           0.648000  2021  
1081           1.095000  2021  
1082           1.205000  2021  
1083           1.895000  2021  

[302 rows x 12 columns]

In [16]:
#calculating Happiness Rank by grouping Happiness Score in descending order 
missing_rank['Happiness Rank'] = missing_rank.groupby('Year')['Happiness Score'].rank('dense',ascending=False)
missing_rank
cut_df.loc[782:1083,'Happiness Rank'] = missing_rank['Happiness Rank']
cut_df[cut_df.Year.isin([2020,2021])]

C:\Users\Agata\AppData\Local\Temp\ipykernel_2852\1278463174.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  missing_rank['Happiness Rank'] = missing_rank.groupby('Year')['Happiness Score'].rank('dense',ascending=False)


Country              Region  Happiness Rank  Happiness Score  \
782       Finland      Western Europe             1.0           7.8087   
783       Denmark      Western Europe             2.0           7.6456   
784   Switzerland      Western Europe             3.0           7.5599   
785       Iceland      Western Europe             4.0           7.5045   
786        Norway      Western Europe             5.0           7.4880   
...           ...                 ...             ...              ...   
1079      Lesotho  Sub-Saharan Africa           143.0           3.5120   
1080     Botswana  Sub-Saharan Africa           144.0           3.4670   
1081       Rwanda  Sub-Saharan Africa           145.0           3.4150   
1082     Zimbabwe  Sub-Saharan Africa           146.0           3.1450   
1083  Afghanistan          South Asia           147.0           2.5230   

      GDP per Capita  Social Support  Healthy Life Expectancy  \
782         1.285190        1.499526                 0.961271   
783         1.326949        1.503449                 0.979333   
784         1.390774        1.472403                 1.040533   
785         1.326502        1.547567                 1.000843   
786         1.424207        1.495173                 1.008072   
...              ...             ...                      ...   
1079        0.451000        0.731000                 0.007000   
1080        1.099000        0.724000                 0.340000   
1081        0.364000        0.202000                 0.407000   
1082        0.457000        0.649000                 0.243000   
1083        0.370000        0.000000                 0.126000   

      Freedom to Make Life Choices  Corruption Perception  Generosity  \
782                       0.662317               0.477857    0.159670   
783                       0.665040               0.495260    0.242793   
784                       0.628954               0.407946    0.269056   
785                       0.661981               0.144541    0.362330   
786                       0.670201               0.434101    0.287985   
...                            ...                    ...         ...   
1079                      0.405000               0.015000    0.103000   
1080                      0.539000               0.088000    0.027000   
1081                      0.627000               0.493000    0.227000   
1082                      0.359000               0.075000    0.157000   
1083                      0.000000               0.010000    0.122000   

      Dystopia Residual  Year  
782            2.762835  2020  
783            2.432741  2020  
784            2.350267  2020  
785            2.460688  2020  
786            2.168266  2020  
...                 ...   ...  
1079           1.800000  2021  
1080           0.648000  2021  
1081           1.095000  2021  
1082           1.205000  2021  
1083           1.895000  2021  

[302 rows x 12 columns]

Happiness Rank is fixed. Let's dive in the Region column to see if the naming is consistent among the years and if not, to pick one year which we'll use as a base for all rows:

In [17]:
#checking if the Region values are the same of different among years
region_naming = cut_df.groupby('Year')['Region'].unique()
pd.set_option('display.max_colwidth', 300)
region_naming

Year
2015         [Western Europe, North America, Australia and New Zealand, Middle East and Northern Africa, Latin America and Caribbean, Southeastern Asia, Central and Eastern Europe, Eastern Asia, Sub-Saharan Africa, Southern Asia]
2016         [Western Europe, North America, Australia and New Zealand, Middle East and Northern Africa, Latin America and Caribbean, Southeastern Asia, Central and Eastern Europe, Eastern Asia, Sub-Saharan Africa, Southern Asia]
2017                                                                                                                                                                                                                            [nan]
2018                                                                                                                                                                                                                            [nan]
2019                                                                       

2015-2016 and 2020-2021 regions are named differenly. 
<br>I chose the first option to be used in my final dataset:

In [18]:
#checking indexing of 2015 - 0:157
cut_df[cut_df.Year.isin([2015])]

Country                           Region  Happiness Rank  \
0    Switzerland                   Western Europe             1.0   
1        Iceland                   Western Europe             2.0   
2        Denmark                   Western Europe             3.0   
3         Norway                   Western Europe             4.0   
4         Canada                    North America             5.0   
..           ...                              ...             ...   
153       Rwanda               Sub-Saharan Africa           154.0   
154        Benin               Sub-Saharan Africa           155.0   
155        Syria  Middle East and Northern Africa           156.0   
156      Burundi               Sub-Saharan Africa           157.0   
157         Togo               Sub-Saharan Africa           158.0   

     Happiness Score  GDP per Capita  Social Support  Healthy Life Expectancy  \
0              7.587         1.39651         1.34951                  0.94143   
1              7.561         1.30232         1.40223                  0.94784   
2              7.527         1.32548         1.36058                  0.87464   
3              7.522         1.45900         1.33095                  0.88521   
4              7.427         1.32629         1.32261                  0.90563   
..               ...             ...             ...                      ...   
153            3.465         0.22208         0.77370                  0.42864   
154            3.340         0.28665         0.35386                  0.31910   
155            3.006         0.66320         0.47489                  0.72193   
156            2.905         0.01530         0.41587                  0.22396   
157            2.839         0.20868         0.13995                  0.28443   

     Freedom to Make Life Choices  Corruption Perception  Generosity  \
0                         0.66557                0.41978     0.29678   
1                         0.62877                0.14145     0.43630   
2                         0.64938                0.48357     0.34139   
3                         0.66973                0.36503     0.34699   
4                         0.63297                0.32957     0.45811   
..                            ...                    ...         ...   
153                       0.59201                0.55191     0.22628   
154                       0.48450                0.08010     0.18260   
155                       0.15684                0.18906     0.47179   
156                       0.11850                0.10062     0.19727   
157                       0.36453                0.10731     0.16681   

     Dystopia Residual  Year  
0              2.51738  2015  
1              2.70201  2015  
2              2.49204  2015  
3              2.46531  2015  
4              2.45176  2015  
..                 ...   ...  
153            0.67042  2015  
154            1.63328  2015  
155            0.32858  2015  
156            1.83302  2015  
157            1.56726  2015  

[158 rows x 12 columns]

In [19]:
c = cut_df.loc[0:157].Country.tolist()
r = cut_df.loc[0:157].Region.tolist()
region = {c[i] : r[i] for i in range(len(c))} #2015 'country : region' dictionary to be used in all years

cut_df['Region'] = cut_df.Country.map(region)
cut_df


Country                           Region  Happiness Rank  \
0     Switzerland                   Western Europe             1.0   
1         Iceland                   Western Europe             2.0   
2         Denmark                   Western Europe             3.0   
3          Norway                   Western Europe             4.0   
4          Canada                    North America             5.0   
...           ...                              ...             ...   
1225     Botswana               Sub-Saharan Africa           142.0   
1226       Rwanda               Sub-Saharan Africa           143.0   
1227     Zimbabwe               Sub-Saharan Africa           144.0   
1228      Lebanon  Middle East and Northern Africa           145.0   
1229  Afghanistan                    Southern Asia           146.0   

      Happiness Score  GDP per Capita  Social Support  \
0               7.587         1.39651         1.34951   
1               7.561         1.30232         1.40223   
2               7.527         1.32548         1.36058   
3               7.522         1.45900         1.33095   
4               7.427         1.32629         1.32261   
...               ...             ...             ...   
1225            3.471         1.50300         0.00100   
1226            3.268         0.00100         0.00000   
1227            2.995         0.00100         0.00100   
1228            2.955         1.39200         0.00000   
1229            2.404         0.00100         0.00000   

      Healthy Life Expectancy  Freedom to Make Life Choices  \
0                     0.94143                       0.66557   
1                     0.94784                       0.62877   
2                     0.87464                       0.64938   
3                     0.88521                       0.66973   
4                     0.90563                       0.63297   
...                       ...                           ...   
1225                  0.28000                       0.57100   
1226                  0.46200                       0.62100   
1227                  0.27000                       0.32900   
1228                  0.63100                       0.10300   
1229                  0.28900                       0.00000   

      Corruption Perception  Generosity  Dystopia Residual  Year  
0                   0.41978     0.29678            2.51738  2015  
1                   0.14145     0.43630            2.70201  2015  
2                   0.48357     0.34139            2.49204  2015  
3                   0.36503     0.34699            2.46531  2015  
4                   0.32957     0.45811            2.45176  2015  
...                     ...         ...                ...   ...  
1225                0.10200     0.01200            0.00000  2022  
1226                0.54400     0.18700            0.00100  2022  
1227                0.10500     0.10600            0.00100  2022  
1228                0.03400     0.08200            0.00000  2022  
1229                0.00500     0.08900            1.26300  2022  

[1230 rows x 12 columns]

There are still countries with missing Region. I checked each of below whether that was caused by different naming or missing region in 2015 year:

In [20]:
cut_df[cut_df['Region'].isna()].Country.unique()

array(['Puerto Rico', 'Belize', 'Somalia', 'Somaliland Region', 'Namibia',
       'South Sudan', 'Taiwan Province of China',
       'Hong Kong S.A.R., China', 'Trinidad & Tobago', 'Northern Cyprus',
       'North Macedonia', 'Gambia', 'Hong Kong S.A.R. of China',
       'Maldives', 'Czechia', 'Congo', 'Eswatini. Kingdom of'],
      dtype=object)

In [21]:
#updating names on countries that had different naming throughout years 
country_correction = {
    'Somaliland Region': 'Somaliland region',
    'Taiwan Province of China': 'Taiwan',
    'Hong Kong S.A.R., China': 'Hong Kong',
    'Trinidad & Tobago': 'Trinidad and Tobago',
    'Northern Cyprus': 'North Cyprus',
    'North Macedonia': 'Macedonia',
    'Hong Kong S.A.R. of China': 'Hong Kong',
    'Czechia': 'Czech Republic'
}
cut_df['Country'] = cut_df.Country.replace(country_correction)

#updating missing region
region_correction = {
    'Puerto Rico': 'Latin America and Caribbean',
    'Belize': 'Latin America and Caribbean',
    'Somalia': 'Sub-Saharan Africa',
    'Namibia': 'Sub-Saharan Africa',
    'South Sudan': 'Sub-Saharan Africa',
    'Gambia': 'Sub-Saharan Africa',
    'Maldives': 'Southern Asia',
    'Congo': 'Sub-Saharan Africa',
    'Eswatini. Kingdom of': 'Sub-Saharan Africa'
}
region.update(region_correction)
cut_df['Region'] = cut_df.Country.map(region)

The dataset looks much better. 
<br>The only null values are Region and Dystopia Residual which means they weren't there in the first place. 

In [22]:
cut_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1230 entries, 0 to 1229
Data columns (total 12 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Country                       1230 non-null   object 
 1   Region                        1230 non-null   object 
 2   Happiness Rank                1230 non-null   float64
 3   Happiness Score               1230 non-null   float64
 4   GDP per Capita                1230 non-null   float64
 5   Social Support                1230 non-null   float64
 6   Healthy Life Expectancy       1230 non-null   float64
 7   Freedom to Make Life Choices  1230 non-null   float64
 8   Corruption Perception         1230 non-null   float64
 9   Generosity                    1230 non-null   float64
 10  Dystopia Residual             918 non-null    float64
 11  Year                          1230 non-null   int32  
dtypes: float64(9), int32(1), object(2)
memory usage: 110.6+ KB


The dataset is ready to be statistically explored.

In [23]:
cut_df.to_csv(r'./cleaned_data.csv', index=False)